integration_test/cases/repo.exs

```diff
@@ -999,4 +999,163 @@ defmodule Ecto.Integration.RepoTest do
     Process.put(:on_log, fn _ -> flunk("logged") end)
     TestRepo.insert!(%Post{title: "1"}, [log: false])
   end
+
+  @tag :upsert
+  test "upsert with struct uses primary key as conflict target" do
+    {:ok, inserted} = TestRepo.upsert(%Post{title: "first"})
+    {:ok, _updated} = TestRepo.upsert(%Post{id: inserted.id, title: "second"})
+
+    post = TestRepo.one!(Post)
+    assert post.title == "second"
+  end
+
+  @tag :upsert
+  test "upsert with struct uses given columns for conflict" do
+    uuid = Ecto.UUID.generate
+
+    {:ok, _} = TestRepo.upsert(%Post{title: "first", uuid: uuid}, conflict_target: [:uuid])
+    {:ok, _} = TestRepo.upsert(%Post{title: "second", uuid: uuid}, conflict_target: [:uuid])
+
+    post = TestRepo.one!(Post)
+    assert post.title == "second"
+  end
+
+  @tag :upsert
+  test "upsert with struct ignored on conflict" do
+    uuid = Ecto.UUID.generate
+
+    {:ok, _} = TestRepo.upsert(%Post{title: "first", uuid: uuid},
+      conflict_target: [:uuid], on_conflict: :ignore)
+    {:ok, _} = TestRepo.upsert(%Post{title: "second", uuid: uuid},
+      conflict_target: [:uuid], on_conflict: :ignore)
+
+    post = TestRepo.one!(Post)
+    assert post.title == "first"
+  end
+
+  @tag :upsert
+  test "upsert with struct increments value on conflict" do
+    uuid = Ecto.UUID.generate
+
+    {:ok, _} = TestRepo.upsert(%Post{title: "first", uuid: uuid, visits: 0},
+      conflict_target: [:uuid], on_conflict: [inc: [visits: 1]])
```

Comment:
Let's add both later. This is good enough to ship as is, then we can add more features later on. :)